compiler/damlc/daml-doc/src/DA/Daml/Doc/Extract.hs

```diff
@@ -119,70 +95,49 @@ extractDocs extractOpts diagsLogger ideOpts fp = do
-collectDocs :: [LHsDecl a] -> [(LHsDecl a, Maybe DocText)]
-collectDocs = go Nothing []
+collectDocs :: [LHsDecl GhcPs] -> [DeclData]
+collectDocs ds
+    | (nextDocs, decl:ds') <- spanMaybe getNextOrPrevDoc ds
+    , (prevDocs, ds'') <- spanMaybe getPrevDoc ds'
+    = DeclData decl (joinDocs nextDocs prevDocs)
+    : collectDocs ds''
+
+    | otherwise
+    = [] -- nothing to document
   where
-    go :: Maybe (LHsDecl p) -> [DocText] -> [LHsDecl p] -> [(LHsDecl p, Maybe DocText)]
-    go Nothing _ [] = []
-    go (Just prev) docs [] = finished prev docs []
-    go prev docs (L _ (DocD _ (DocCommentNext str)) : ds)
-      | Nothing <- prev = go Nothing (docToText str:docs) ds
-      | Just decl <- prev = finished decl docs (go Nothing [docToText str] ds)
-    go prev docs (L _ (DocD _ (DocCommentPrev str)) : ds) =
-      go prev (docToText str:docs) ds
-    go Nothing docs (d:ds) = go (Just d) docs ds
-    go (Just prev) docs (d:ds) = finished prev docs (go (Just d) [] ds)
-
-    finished decl docs rest = (decl, toDocText . map unDocText . reverse $ docs) : rest
-
--- | Context in which to extract a module's docs. This is created from
--- 'TypecheckedModule' by 'buildDocCtx'.
-data DocCtx = DocCtx
-    { dc_ghcMod :: GHC.Module
-        -- ^ ghc name for current module
-    , dc_modname :: Modulename
-        -- ^ name of the current module
-    , dc_tcmod :: TypecheckedModule
-        -- ^ typechecked module
-    , dc_decls :: [DeclData]
-        -- ^ module declarations
-    , dc_insts :: [ClsInst]
-        -- ^ typeclass instances
-    , dc_tycons :: MS.Map Typename TyCon
-        -- ^ types defined in this module
-    , dc_datacons :: MS.Map Typename DataCon
-        -- ^ constructors defined in this module
-    , dc_ids :: MS.Map Fieldname Id
-        -- ^ values defined in this module
-    , dc_templates :: Set.Set Typename
-        -- ^ DAML templates defined in this module
-    , dc_choices :: MS.Map Typename (Set.Set Typename)
-        -- ^ choices per DAML template defined in this module
-    , dc_extractOptions :: ExtractOptions
-        -- ^ command line options that affect the doc extractor
-    , dc_exports :: ExportSet
-        -- ^ set of export, unless everything is exported
-    }
-
--- | Parsed declaration with associated docs.
-data DeclData = DeclData
-    { _dd_decl :: LHsDecl GhcPs
-    , _dd_docs :: Maybe DocText
-    }
+
+    joinDocs :: [DocText] -> [DocText] -> Maybe DocText
+    joinDocs nextDocs prevDocs =
+        let docs = map unDocText (nextDocs ++ prevDocs)
+        in if null docs
+            then Nothing
+            else Just . DocText . T.strip $ T.unlines docs
+
+    getNextOrPrevDoc :: LHsDecl a -> Maybe DocText
+    getNextOrPrevDoc = \case
+        L _ (DocD _ (DocCommentNext str)) -> Just (docToText str)
+        L _ (DocD _ (DocCommentPrev str)) -> Just (docToText str)
+            -- technically this is a malformed doc, but we'll take it
```

Comment:
I initially misunderstood the line this was referring to. Maybe add it on the same line or make it clear in the comment that this is about `DocCommentPrev`.

Reply:
I added a little `^`, hopefully that makes it clearer :)

Reply:
thanks!